### PR TITLE
Show pprof labels in thread names

### LINKED ISSUE #3493: dap: show pprof labels in thread names
It would be useful to see pprof labels in VS Code's debugger. One possible way to show them is to append them after the go routine ID in the thread names, e.g. instead of showing just `[Go 42]`, show `[Go 42 key1: value1, key2: value2]`.

image

1. What version of Delve are you using (`dlv version`)? **1.20.2**
2. What version of Go are you using? (`go version`)? **go1.20.7 darwin/arm64**
3. What operating system and processor architecture are you using? **MacOS 13.5.1, M1**

### PR DESCRIPTION
**Pull request by stefanhaller**
If a goroutine has labels, we add them after the ID.

Fixes #3493.

### CODE REVIEW

File: service/dap/server.go [hunk 1]
```diff
@@ -1800,10 +1800,21 @@ func (s *Session) onThreadsRequest(request *dap.ThreadsRequest) {
...
 			}
+			labels := ""
+			if len(g.Labels()) > 0 {
+				var keys []string
+				for k := range g.Labels() {
+					keys = append(keys, k)
+				}
+				sort.Strings(keys)
+				for _, k := range keys {
+					labels += fmt.Sprintf(" %s=%s", k, g.Labels()[k])
```
Comment (aarzilli) [inline service/dap/server.go:1811]: use a `strings.Builder`

**Review by aarzilli — COMMENTED** [review]
cc @hyangah @suzmue

File: service/dap/server.go (hunk 1 above)
Comment (stefanhaller) [reply service/dap/server.go]: Ah yes, thanks. See 93d379b059a2.

**Review by aarzilli — APPROVED** [review]

**Comment by stefanhaller** [comment]
After testing this for a while I found that `k1:v1` looks better than `k1=v1`, so I pushed a fixup (515f4cdc09bc) to change that. Not a big thing, let me know if you disagree.

**Comment by derekparker** [comment]
@hyangah or @suzmue would you mind taking a look at this?

**Comment by stefanhaller** [comment]
@derekparker Please note that over in #3493 some concerns were brought up; showing all labels might be too much when there are many. Please see the discussion there, starting at https://github.com/go-delve/delve/issues/3493#issuecomment-1734352570.

**Comment by suzmue** [comment]
I think this looks great but would love to see this behind an opt-in configuration to start! (commented on #3493 as well)

File: service/dap/server.go [hunk 2]
```diff
@@ -1800,10 +1800,21 @@ func (s *Session) onThreadsRequest(request *dap.ThreadsRequest) {
...
+			if len(g.Labels()) > 0 {
+				var keys []string
+				for k := range g.Labels() {
+					keys = append(keys, k)
+				}
+				sort.Strings(keys)
+				for _, k := range keys {
+					fmt.Fprintf(&labels, " %s:%s", k, g.Labels()[k])
+				}
+			}
```
Comment (alexandear) [inline service/dap/server.go:1813]: Can we rewrite it as follows:
```suggestion
			keys := make([]string, 0, len(g.Labels()))
			for k := range g.Labels() {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			var labels strings.Builder
			for _, k := range keys {
				labels.WriteByte(' ')
				labels.WriteString(k)
				labels.WriteByte(':')
				labels.WriteString(g.Labels()[k])
			}
```
?

**Comment by stefanhaller** [comment]
@suzmue I force-pushed to implement the config, please have a look!

File: service/dap/server.go (hunk 2 above)
Comment (stefanhaller) [reply service/dap/server.go]: The implementation has changed after the discussion in #3493, as we don't show all labels any more.

I'm curious though, why is writing individual bytes manually better than Fprintf to a builder?

**Comment by suzmue** [comment]
I think having a more general solution here could be better. I think having the flexibility to look at all the pprof labels would be great.

We could combine that with another config to only show specific labels but I don't think that needs to go into this change, but I either a list of labels to show or a string that can be parsed like goroutineFilters would be good. It would be nice to be able to show both a single label like the name use case, and a combination of labels that are interesting to the user.

We can also make the pprof labels config easier to adjust / more discoverable by adding it to the callstack context menu in VS Code. Once we get this change in here I am happy to add it (see [hide system goroutines example](https://github.com/golang/vscode-go/issues/1797#issuecomment-1762145244))

**Comment by stefanhaller** [comment]
@suzmue Now I'm confused. 😅 Could you please be more specific as to what you would like me to do in this PR?

It sounds like you are thinking of having two different configs, where the second could be added later and would be some sort of filter for the labels. But then what's the first one? A bool option to enable the feature? But once we add the filter, would we still need the bool option?

I'm happy to implement the second option now, whatever it is, but I'd like to have agreement over the exact design before I start, in order to avoid wasting time. My proposal would be to add a single config, `showPprofLabels`, which is a list of strings, default empty. No globbing or regex, just a simple list of keys, because that simplifies things a lot. As a special case for my use case, I'd show just the value of the label if the length of the array is one, and `k:v` otherwise. The downside of this approach is that there's no easy way to say "show all keys" without listing them all manually; but we could add the special case of a single element of `"*"` for that, if needed.

Thoughts?

**Comment by suzmue** [comment]
@stefanhaller I think your idea of showPprofLabels as a list of strings sounds great! Those two special cases also sound good to me. Thanks!

**Comment by stefanhaller** [comment]
@suzmue I force pushed to implement the proposed solution.

**Comment by stefanhaller** [comment]
Ping. @suzmue, would you have time to have a look at this? (Or anybody else?)

File: service/dap/config.go [hunk 3]
```diff
@@ -85,3 +94,37 @@ func configureSetSubstitutePath(args *launchAttachArgs, rest string) error {
 	}
 	return nil
 }
+
+func configureSetShowPprofLabels(args *launchAttachArgs, rest string) error {
```
Comment (suzmue) [inline service/dap/config.go:98]: Could you add documentation to command.go explaining how this configuration works please? For msgConfig

**Review by suzmue — CHANGES_REQUESTED** [review]

File: service/dap/server.go [hunk 4]
```diff
@@ -1812,10 +1822,41 @@ func (s *Session) onThreadsRequest(request *dap.ThreadsRequest) {
 			if g.Thread != nil && g.Thread.ThreadID() != 0 {
 				thread = fmt.Sprintf(" (Thread %d)", g.Thread.ThreadID())
 			}
+			var labels strings.Builder
+			appendLabelsForKeys := func(keys []string) {
```
[the remark this answers is not in the document]
Comment (stefanhaller) [reply service/dap/server.go]: Renamed in 49a08989663e

File: service/dap/server_test.go [hunk 5]
```diff
@@ -1386,10 +1393,94 @@ func TestSelectedThreadsRequest(t *testing.T) {
...
+				// Breakpoints are set within the program
+				"", []int{},
+				[]onBreakpoint{{
+					execute: func() {
+						client.ThreadsRequest()
+						tr := client.ExpectThreadsResponse(t)
+						if len(tr.Body.Threads) != 1 {
+							t.Errorf("got %d threads, expected 1\n", len(tr.Body.Threads))
+						}
+						// The first breakpoint is before the call to pprof.Do; no lables yet:
```
[the remark this answers is not in the document]
Comment (stefanhaller) [reply service/dap/server_test.go]: Fixed in fb5bcaf9234e

File: service/dap/server_test.go [hunk 6]
```diff
@@ -1386,10 +1393,94 @@ func TestSelectedThreadsRequest(t *testing.T) {
...
-
 	})
 }
 
+func TestGoroutineLabels(t *testing.T) {
+	tests := []struct {
+		showPprofLabelsConfig   []string
+		expectedPrefixWithLabel string
+	}{
+		{
```
[the remark this answers is not in the document]
Comment (stefanhaller) [reply service/dap/server_test.go]: Sure, reformatted in d5ac1c646505

File: service/dap/config.go (hunk 3 above)
Comment (stefanhaller) [reply service/dap/config.go:98]: Added in 285116f8256e

**Comment by stefanhaller** [comment]
Thanks for the review, I addressed all requests. I referenced the individual fixup commits above to make it easier to see what I changed, but I'm going to squash them now (and rebase on master again).

**Review by hyangah — APPROVED** [review]

**Pull request merged**